bee-network/bee-autopeering/src/peer/mod.rs

```diff
@@ -0,0 +1,334 @@
+// Copyright 2021 IOTA Stiftung
+// SPDX-License-Identifier: Apache-2.0
+
+pub(crate) mod lists;
+
+pub mod peer_id;
+pub mod stores;
+
+pub use peer_id::PeerId;
+pub use stores::PeerStore;
+
+use lists::{ActivePeersList, ReplacementList};
+
+use crate::{
+    local::{
+        services::{ServiceMap, ServiceProtocol},
+        Local,
+    },
+    proto,
+};
+
+use bytes::BytesMut;
+use crypto::signatures::ed25519::PublicKey;
+use libp2p_core::{multiaddr::Protocol, Multiaddr};
+use prost::{DecodeError, EncodeError, Message};
+use serde::{
+    de::{SeqAccess, Visitor},
+    ser::SerializeStruct,
+    Deserialize, Serialize,
+};
+
+use std::{fmt, net::IpAddr};
+
+/// Represents a peer.
+#[derive(Clone)]
+pub struct Peer {
+    peer_id: PeerId,
+    ip_address: IpAddr,
+    services: ServiceMap,
+}
+
+impl Peer {
+    /// Creates a new instance.
+    pub fn new(address: IpAddr, public_key: PublicKey) -> Self {
+        let peer_id = PeerId::from_public_key(public_key);
+
+        Self {
+            peer_id,
+            ip_address: address,
+            services: ServiceMap::default(),
+        }
+    }
+
+    /// Returns the [`PeerId`] of this peer.
+    pub fn peer_id(&self) -> &PeerId {
+        &self.peer_id
+    }
+
+    /// Returns the public key of this peer.
+    pub fn public_key(&self) -> &PublicKey {
+        self.peer_id.public_key()
+    }
+
+    /// Returns the address of this peer.
+    pub fn ip_address(&self) -> IpAddr {
+        self.ip_address
+    }
+
+    /// Returns the port of a service provided by this peer.
+    pub fn port(&self, service_name: impl AsRef<str>) -> Option<u16> {
+        self.services().get(service_name).map(|s| s.port())
+    }
+
+    /// Returns the services of this peer.
+    pub fn services(&self) -> &ServiceMap {
+        &self.services
+    }
+
+    /// Sets the services of this peer all at once.
+    pub(crate) fn set_services(&mut self, services: ServiceMap) {
+        self.services = services;
+    }
+
+    /// Returns whether the peer provides a corresponding service.
+    pub fn has_service(&self, service_name: impl AsRef<str>) -> bool {
+        self.services.get(service_name).is_some()
+    }
+
+    /// Adds a service with address binding to this peer.
+    pub fn add_service(&mut self, service_name: impl ToString, protocol: ServiceProtocol, port: u16) {
+        self.services.insert(service_name.to_string(), protocol, port);
+    }
+
+    /// Returns the address, i.e. the `libp2p_core::Multiaddr`, belonging to the given service name.
+    ///
+    /// Example: "peering" => `/ip4/127.0.0.1/udp/14627`.
+    pub fn get_service_multiaddr(&self, service_name: impl AsRef<str>) -> Option<Multiaddr> {
```

Comment:
Tiny bit inconsistent with the `port` method, they do something very similar but are called very differently

Reply:
you're right

Reply:
I like it short so I'll rename it to just `multiaddr`.

Reply:
Done